scripts/esbuild/cli.ts

```diff
@@ -0,0 +1,74 @@
+import type { BuildOptions as ESBuildOptions } from 'esbuild';
+import * as esbuild from 'esbuild';
+import fs from 'fs-extra';
+import { join } from 'path';
+
+import { BuildOptions } from '../utils/options';
+import { writePkgJson } from '../utils/write-pkg-json';
+import { getEsbuildAliases, getEsbuildExternalModules } from './util';
+
+/**
+ * Runs esbuild to bundle the `cli` submodule
+ *
+ * @param opts build options
+ */
+export async function buildCli(opts: BuildOptions) {
+  // clear out rollup stuff
+  fs.emptyDir(opts.output.cliDir);
+
+  const inputDir = join(opts.srcDir, 'cli');
+  const buildDir = join(opts.buildDir, 'cli');
+
+  const outputDir = opts.output.cliDir;
+  const esmFilename = 'index.js';
+  const cjsFilename = 'index.cjs';
+
+  const dtsFilename = 'index.d.ts';
+
+  const cliAliases = getEsbuildAliases(opts);
+
+  const external = getEsbuildExternalModules(opts, opts.output.cliDir);
+
+  const cliEsbuildOptions: ESBuildOptions = {
```

Comment:
Is there an option that is equivalent to Rollup's `preferConst`? It looks like entries get compiled down to `var`s:
![Screenshot 2023-10-10 at 1 31 32 PM](https://github.com/ionic-team/stencil/assets/1930213/aeb508a2-9939-41df-aa50-23d14d8914db)

Reply:
I did a little googling around and it seems that this is not something that esbuild supports, and for some of what seem like decent reasons. They have an FAQ entry about it here: <https://esbuild.github.io/faq/#top-level-var>

The tl;dr as I understand it is that esbuild will sometimes need to lazily initialize an imported module[^1] and for performance reasons when it does so it will separate variable declaration from initialization so it of course can't use `const` to declare those variables since initialization has to happen at variable declaration with `const`. Then I think esbuild prefers `var` over `let` because of those TDZ-check related performance issues.

I don't really know enough about the mechanics of bundling or about JS VM performance to fully assess the validity of the argument, but that's what they're saying as to why there's no equivalent of a `preferConst` option.

[^1]: I think especially with an `import()` call but I couldn't _quite_ get this to happen in a small example so I'm not sure exactly what they're referring to